### PR TITLE
Re-implement pin_memory to be device-agnostic by leveraging the Accelerator concept

### PR DESCRIPTION
**Pull request by wizzniu**
This PR re-implements pin memory aiming to get rid of the optional `device` argument and makes all related APIs to be device-agnostic. We add two new abstract APIs in [AcceleratorHooksInterface](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/detail/AcceleratorHooksInterface.h#L12) and redefine pin memory as: "Pin memory is always pinned for the current accelerator device". In detail, it uses [getAcceleratorHooksInterface](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/Context.h#L61) in pin_memory/is_pinned to get an appropriate device and invoke the corresponding overridden interfaces, instead of using BackendSelect and then dispatching to CUDA or other specific backends' implement methods.

Note: For new backends who want to implement and use pin memory, just inherit AcceleratorHooksInterface and overwrite the `isPinnedPtr` and `getPinnedMemoryAllocator` methods. 

Additional context: To avoid BC-breaking, this PR just preserves the `device` arg of related APIs and would throw a deprecation warning if `device` arg is passed. Another PR will be submitted to update all PT callers (`Tensor.is_pinned()`, `Tensor.pin_memory()`...) not to pass this arg based on this PR. In future, `device` arg will be actually removed.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @LucasLLC @MeetVadakkanchery @mhorowitz @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l @albanD @ezyang 

Relates #124908
Relates #14560

### CODE REVIEW

**Comment by wizzniu** [comment]
@albanD @ezyang Could you help to review this? If it's reasonable, I will go on for the next step to refresh all related APIs and modify the test cases.

File: aten/src/ATen/native/Memory.cpp
Comment (legionGIT) [inline aten/src/ATen/native/Memory.cpp:40]: I think it's better to use _pin_memory here.

Comment (wizzniu) [reply aten/src/ATen/native/Memory.cpp:40]: Since I noticed that the different backends' implement of `_pin_memory` is mainly at host side and  the code's logic is common, where the main difference is in how to get their own Allocator. So I remove function `_pin_memory` and don't go a redispatching way.

**Comment by ezyang** [comment]
@albanD for you

File: aten/src/ATen/mps/MPSPinnedMemory.h
Comment (albanD) [inline aten/src/ATen/mps/MPSPinnedMemory.h:7]: Do we need a new file for this, can this be placed next to getIMPSAllocator() ?

File: aten/src/ATen/native/Memory.cpp
Comment (albanD) [reply aten/src/ATen/native/Memory.cpp:40]: Hey!
I'm afraid you're going to have to keep `_pin_memory` and the reason is from what is happening in native_functions.yaml.

File: aten/src/ATen/native/native_functions.yaml
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:4534]: This function's implementation is NOT Composite. So we have to add a CompositeExplicitAutograd: pin_memory here.
This will lead to the fact that you must add back the derivatives formula and the aliasing information must be accurate.

Since this function sometimes return a view and sometimes not, you won't be able to have the aliasing info correctly and will have to do the same trick as before:
pin_memory remain compositeimplicit (no specific dispatch for it) and can have inncurate aliasing info as before.
You have a _pin_memory that is never aliasing that is compositeexplicitautograd that is accurate and has a derivatives.yaml formula.

tl;dr: you have to restore the _pin_memory trick I'm afraid to have proper autograd/aliasing behavior.
Also since you have to keep that, you might want to keep the _pin_memory_nested as well on top of the compositeexplicitimplementation

File: aten/src/ATen/native/native_functions.yaml
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:4529]: This is going to be badly BC-breaking I'm afraid.
To make this PR landing more easily, I would preserve the device arg here and continue to respect it. Only change the default to point to the accelerator.

We can then have 3 follow ups:
- Update all pt callers not to pass this arg
- Throw a deprecation warning if this arg is passed
- After 2 releases, actually remove the arg.

File: aten/src/ATen/detail/AcceleratorHooksInterface.h
Comment (wizzniu) [reply aten/src/ATen/detail/AcceleratorHooksInterface.h]: I think both is okay. Have used named argument instead, also for all subclasses.

File: aten/src/ATen/mps/MPSPinnedMemory.h
Comment (wizzniu) [reply aten/src/ATen/mps/MPSPinnedMemory.h]: It's really not necessary. Have removed this file and place it next to getIMPSAllocator() in `aten/src/ATen/mps/MPSAllocatorInterface.h`.

File: aten/src/ATen/native/native_functions.yaml
Comment (wizzniu) [reply aten/src/ATen/native/native_functions.yaml]: Yes, you're right. I ignored the aliasing information before. And it's difficult to use CompositeExplicitAutograd: pin_memory here.
So I have restored the _pin_memory trick and use CompositeImplicitAutograd: pin_memory to  guarantee proper autograd behavior.
Also for _pin_memory, keep the _pin_memory_nested on top of the CompositeExplicitAutograd. Here, I only reserve NestedTensorCPU key, cause manual backendselect for _pin_memory is removed and input tensor `self` can be performed pin_memory operation only if it's a cpu tensor. So unlike before, NestedTensorCUDA key and other nested keys are just meaningless for _pin_memory now. When inputting a nested cuda tensor, the behavior changes from throwing a RuntimeError before("only dense CPU tensors can be pinned") to throwing a NotImplementedError now ("Could not run aten::_pin_memory with arguments from the NestedTensorCUDA backend").

File: aten/src/ATen/native/native_functions.yaml
Comment (wizzniu) [reply aten/src/ATen/native/native_functions.yaml]: Agree with you.

- If low level implement of pin_memory is acceptable, I will go on for the next step to update all PT callers of related APIs.
- Have added deprecation warnings in C++ side. But I'm not sure if it's enough, WDYT?

File: aten/src/ATen/native/Memory.cpp
Comment (wizzniu) [reply aten/src/ATen/native/Memory.cpp:58]: ok, have added it back.

File: aten/src/ATen/native/native_functions.yaml
Comment (albanD) [reply aten/src/ATen/native/native_functions.yaml]: Hey,
Will take a look at the code.
By what I meant by 3 follow ups here meant that this can be 3 different PRs (doesn't have to be).

Comment (wizzniu) [reply aten/src/ATen/native/native_functions.yaml]: Thanks!
As what you said, to make this PR landing more easily, this PR will only refactor implement of pin memory and add a deprecation warning for `device` arg. I will submit another PR to update all PT callers based on this PR.

File: aten/src/ATen/detail/HIPHooksInterface.h
Comment (albanD) [inline aten/src/ATen/detail/HIPHooksInterface.h:25]: FYI @jeffdaily this makes HIP an "accelerator". I'm still not sure if you use it but that's just enabling more device-generic feature for the HIP device so I guess you're happy with it. We can remove it if you are not!

File: aten/src/ATen/detail/MTIAHooksInterface.h
Comment (albanD) [inline aten/src/ATen/detail/MTIAHooksInterface.h:98]: FYI @egienvalue you can implement this if/when you need to support pinned host-side memory used for faster transfers to the device!

**Review by albanD — DISMISSED** [review]
Sounds good!
This needs rebasing on latest main so that we get CI signal btw!

File: aten/src/ATen/detail/HIPHooksInterface.h
Comment (jeffdaily) [reply aten/src/ATen/detail/HIPHooksInterface.h:25]: IIRC we don't use HIPHooksInterface but rather a hipified version of the CUDAHooksInterface.  In any case, we're okay with being an Accelerator.  Is there an RFC or something similar describing pytorch's move to these generic interfaces?

Comment (albanD) [reply aten/src/ATen/detail/HIPHooksInterface.h:25]: The short version is described in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/DeviceAccelerator.h

**Comment by wizzniu** [comment]
@albanD Have rebased. Can we merge it now?

File: aten/src/ATen/native/native_functions.yaml
Comment (albanD) [inline aten/src/ATen/native/native_functions.yaml:4535]: Ho sorry for the oversight on my end.
The implementation here is not a valid CompositeImplicitAutograd implementation (the default key when nothing is specified as it calls into non-aten ops (the context).

Please update to
```

  dispatch:
    CompositeExplicitAutograd: is_pinned
```

File: aten/src/ATen/detail/MTIAHooksInterface.h
Comment (egienvalue) [reply aten/src/ATen/detail/MTIAHooksInterface.h:98]: That is nice. Right now we had a hacky way to pin CPU tensors.

File: aten/src/ATen/native/native_functions.yaml
Comment (wizzniu) [reply aten/src/ATen/native/native_functions.yaml]: Yes, updated.

**Comment by wizzniu** [comment]
@albanD Are there any other problems? Maybe we can start CI to see if any problem exists.

**Comment by wizzniu** [comment]
Hi, @albanD I have pushed new commit to handle the failed test case. Can we trigger ci now? It seems that I don't have permissions.

File: aten/src/ATen/detail/MPSHooksInterface.h
Comment (kulinseth) [inline aten/src/ATen/detail/MPSHooksInterface.h:95]: @wizzniu and @albanD 
What is needed to adopt the MPS to the accelerator interface ? If there are APIs missing, we would like to extend it.

Comment (wizzniu) [reply aten/src/ATen/detail/MPSHooksInterface.h:95]: @kulinseth Seems that there is no need to add extra APIs for MPS currently. But I don't test mps locally. It still need ci test for mps to verify.

**Comment by wizzniu** [comment]
@albanD Do you have time to look at the newly pushed commits? 
And @ezyang ,Could you help to review the code,  especially for the part of op's registration and dispatch? ( I notice that you are the code owner of pin memory

**Comment by ezyang** [comment]
It's probably best if @albanD finishes off the review here

**Comment by wizzniu** [comment]
> It's probably best if @albanD finishes off the review here

Thanks! Seems that he is busy.

**Review by albanD — DISMISSED** [review]
Thanks for the update! That looks great!

**Comment by LucasLLC** [comment]
We recently ran into an issue where calling `is_pinned` on a cpu tensor would initiate a cuda context if device was != cpu. I'm curious what the expected behavior is since merging this PR?

cc: @wizzniu @albanD

**Comment by albanD** [comment]
@LucasLLC can you open an issue with a repro for this please? This is not expected for sure but we can track that independently of this PR.

**Comment by LucasLLC** [comment]
@albanD Should have included in previous comment, here is the issue: https://github.com/pytorch/pytorch/issues/128988

**Comment by daulet-askarov** [comment]
> @daulet-askarov has imported this pull request. If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59255760).

cc: @LucasLLC, I just tried import this PR and on top of it dropped the device param to our is_pinned calls in DCP. The staging memory consumption test still passes, so looks like we should be okay reverting PR128896 after the current PR gets merged.

**Comment by wizzniu** [comment]
> We recently ran into an issue where calling `is_pinned` on a cpu tensor would initiate a cuda context if device was != cpu. I'm curious what the expected behavior is since merging this PR?
> 
> cc: @wizzniu @albanD

- Only cpu tensor can be pinned, the `device` that input tensor pinned memory on is not up to itself.
- Currently, CUDA is the default device to pin memory if `device` arg is not passed. After merging this PR, we change the default device from CUDA to auto-selected accelerator(not including CPU).
- So whether initiating the cuda context when calling `is_pinned` depends on the detailed implement of [isPinnedPtr](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cuda/detail/CUDAHooks.cpp#L124). And this PR doesn't modify the internal implement of `CUDAHooks::isPinnedPtr`.
- It's better to open another independent issue to track it.

**Comment by wizzniu** [comment]
> > @daulet-askarov has imported this pull request. If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59255760).
> 
> cc: @LucasLLC, I just tried import this PR and on top of it dropped the device param to our is_pinned calls in DCP. The staging memory consumption test still passes, so looks like we should be okay reverting PR128896 after the current PR gets merged.

As mentioned above, this PR aims to remove the `device` arg of `is_pinned` and `pin_memory`. I will refresh APIs in torch to not passing `device` arg. Now in this PR, it's not supported to pass `device='cpu'` in `is_pinned`. So I think PR128896 should be reverted first before this PR getting merged.
And another thing to confirm, @albanD Previously, it's allowed to pass `device='cpu'` in `is_pinned` and it will return false. But in this PR, it's not allowed and will give an error 'CPU device type not an accelerator'. So whether it should be allowed to pass `device='cpu'` for BC?

**Comment by guangyey** [comment]
Hi @wizzniu, this PR has a conflict with https://github.com/pytorch/pytorch/pull/129463 and https://github.com/pytorch/pytorch/pull/129205. Please make sure to note to rebase to resolve the conflicts. Sorry for your inconvenience.

**Comment by daulet-askarov** [comment]
@wizzniu, okay, PR 129972 has been merged, this should fix the distributed/checkpoint/test_utils. Can you please try to re-base and merge?

File: aten/src/ATen/native/Memory.cpp
Comment (albanD) [inline aten/src/ATen/native/Memory.cpp:27]: As you mentioned, we definitely want to preserve this behavior.
In this case, I think we need to properly handle the case where there is no accelerator or the passed device is not an accelerator and always return "false" for these.
That should preserve the old behavior for non-accelerator devices right?

Comment (wizzniu) [reply aten/src/ATen/native/Memory.cpp:27]: @albanD Yes, updated.

**Comment by wizzniu** [comment]
@albanD Is there any problem? Maybe we can merge it now.

**Review by albanD — DISMISSED** [review]
Looks good, thanks!

**Comment by wizzniu** [comment]
@guangyey Can you help to look at this xpu's compilation failure? Seems unrelated to this PR, maybe need rebase first?

**Comment by guangyey** [comment]
"unrelated failures"
@pytorchbot merge -i

**Comment by wizzniu** [comment]
"unrelated failures"
@pytorchbot merge -i

**Comment by guangyey** [comment]
> "unrelated failures" @pytorchbot merge -i
I will help you merge this PR if CI is happy.

**Comment by guangyey** [comment]
> test_inductor_counters

Unrelated to this PR.

**Comment by wizzniu** [comment]
> @wizzniu test_inductor_counters failure, related?

Seems unrelated. Caused by https://github.com/pytorch/pytorch/commit/156b99cfb1c90a43e588168d6e18d1c9737525ad

**Comment by cyyever** [comment]
@wizzniu Add ROCM tests. Once they are passed, I will merge it.

**Comment by atalman** [comment]
Error during build:
```
aten/src/ATen/Context.h:101:21: error: reference to 'DeviceType' is ambiguous
   ->       std::optional<DeviceType> device_type = std::nullopt) {
   ->                     ^
   -> xplat/caffe2/c10/core/DeviceType.h:36:12: note: candidate found by name lookup is 'c10::DeviceType'
   -> enum class DeviceType : int8_t {
   ->            ^
```

File: aten/src/ATen/Context.h
Comment (albanD) [inline aten/src/ATen/Context.h:101]: Can you use `c10::DeviceType` here and below to avoid that issue?

**Comment by albanD** [comment]
@atalman could we track that internally to see why there is divergence between OSS and Internal signal?

File: aten/src/ATen/Context.h
Comment (wizzniu) [reply aten/src/ATen/Context.h]: Have updated.

File: aten/src/ATen/DeviceAccelerator.cpp
Comment (albanD) [inline aten/src/ATen/DeviceAccelerator.cpp:5]: Why did you remove the C10_API here?

**Review by albanD — APPROVED** [review]
The removal of C10_API might cause issues for windows build. But if that passes CI, I'm happy!

File: aten/src/ATen/DeviceAccelerator.cpp
Comment (wizzniu) [reply aten/src/ATen/DeviceAccelerator.cpp:5]: Since there exists  [TORCH_API](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/DeviceAccelerator.h#L25) in header file. I just thought it was a mistake. Will add it back if it causes issues.

**Pull request closed without merging**
